### PR TITLE
Improve outbound governor memory footprint

### PR DESCRIPTION
**Pull request by coot**
# Description

I run `cabal run ouroboros-network:sim-tests -- -p '/PeerSelection.progresses/' +RTS -t -RTS`
optionally with `-N4`. Below are the results.

# Unoptimised
```
<<ghc: 35570273736 bytes, 8606 GCs, 823497/2268168 avg/max bytes residency (690 samples), 11M in use, 0.001 INIT (0.000 elapsed), 11.840 MUT (11.824 elapsed), 1.364 GC (1.371 elapsed) :ghc>>
<<ghc: 35569891816 bytes, 8606 GCs, 833172/2822640 avg/max bytes residency (674 samples), 11M in use, 0.001 INIT (0.000 elapsed), 11.591 MUT (11.568 elapsed), 1.351 GC (1.358 elapsed) :ghc>>
<<ghc: 35570152208 bytes, 8606 GCs, 839346/2894024 avg/max bytes residency (688 samples), 12M in use, 0.001 INIT (0.001 elapsed), 12.092 MUT (12.075 elapsed), 1.469 GC (1.475 elapsed) :ghc>>
```
## `-N4`
```
<<ghc: 35566896400 bytes, 2529 GCs, 2069056/7172688 avg/max bytes residency (307 samples), 39M in use, 0.002 INIT (0.001 elapsed), 15.837 MUT (4.226 elapsed), 2.079 GC (0.805 elapsed) :ghc>>
<<ghc: 35567056840 bytes, 2488 GCs, 2014886/7764896 avg/max bytes residency (310 samples), 38M in use, 0.001 INIT (0.001 elapsed), 15.835 MUT (4.201 elapsed), 2.061 GC (0.799 elapsed) :ghc>>
<<ghc: 35567023304 bytes, 2492 GCs, 2032549/6684632 avg/max bytes residency (306 samples), 39M in use, 0.001 INIT (0.001 elapsed), 15.788 MUT (4.187 elapsed), 2.032 GC (0.791 elapsed) :ghc>>
```
# Optimised
```
<<ghc: 35694618112 bytes, 8638 GCs, 830121/2715632 avg/max bytes residency (666 samples), 11M in use, 0.000 INIT (0.000 elapsed), 12.274 MUT (12.256 elapsed), 1.452 GC (1.459 elapsed) :ghc>>
<<ghc: 35694541536 bytes, 8638 GCs, 825957/2724136 avg/max bytes residency (668 samples), 11M in use, 0.000 INIT (0.000 elapsed), 12.421 MUT (12.406 elapsed), 1.419 GC (1.426 elapsed) :ghc>>
<<ghc: 35694425144 bytes, 8638 GCs, 819758/2531536 avg/max bytes residency (671 samples), 11M in use, 0.001 INIT (0.000 elapsed), 11.854 MUT (11.838 elapsed), 1.397 GC (1.404 elapsed) :ghc>>
```
## `-N4`
```
<<ghc: 35691456424 bytes, 2513 GCs, 1932057/7059928 avg/max bytes residency (312 samples), 36M in use, 0.001 INIT (0.001 elapsed), 15.870 MUT (4.214 elapsed), 2.025 GC (0.802 elapsed) :ghc>>
<<ghc: 35691356248 bytes, 2502 GCs, 2008709/7795608 avg/max bytes residency (301 samples), 38M in use, 0.001 INIT (0.001 elapsed), 15.869 MUT (4.215 elapsed), 2.068 GC (0.817 elapsed) :ghc>>
<<ghc: 35691420256 bytes, 2515 GCs, 1989430/6552568 avg/max bytes residency (305 samples), 39M in use, 0.002 INIT (0.001 elapsed), 15.846 MUT (4.203 elapsed), 2.028 GC (0.800 elapsed) :ghc>>
```
# Checklist

- Branch
    - [ ] Updated changelog files.
    - [ ] Commit sequence broadly makes sense
    - [ ] Commits have useful messages
    - [ ] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

**Comment by coot** [comment]
I have more interesting improvements & results.  These include yet unpushed changes to `io-sim`'s `stm` operations.  But it shows that the changes in this branch actually make a difference:

# IOSim Memory Profiling

### Optimised IOSim, Optimised O-N

* `io-sim` (OPT): `3be6e9b8ffe8ba75ad88c7481cf846c9e90e85cf`
* `ouroboros-network` (OPT): `a7fd4783a258ea10912ea2d6ea37a74103c0f5bf`

```
cabal run ouroboros-network:sim-tests -- -p '/progress.public root peers.progresses towards established peers/' +RTS -t -RTS --quickcheck-replay=824041
```

```
<<ghc: 71952108616 bytes, 13359 GCs, 10128437/70286680 avg/max bytes residency (28 samples), 208M in use, 0.000 INIT (0.000 elapsed), 11.746 MUT (11.729 elapsed), 1.419 GC (1.429 elapsed) :ghc>>
<<ghc: 71952127960 bytes, 13359 GCs, 11075894/70722376 avg/max bytes residency (29 samples), 209M in use, 0.000 INIT (0.000 elapsed), 11.622 MUT (11.599 elapsed), 1.464 GC (1.473 elapsed) :ghc>>
<<ghc: 71952118040 bytes, 13359 GCs, 12941755/77539016 avg/max bytes residency (30 samples), 228M in use, 0.001 INIT (0.000 elapsed), 11.129 MUT (11.111 elapsed), 1.559 GC (1.570 elapsed) :ghc>>
```

### Unoptimised IOSim, Optimised O-N


* `io-sim` (UNOPT): `3be6e9b8ffe8ba75ad88c7481cf846c9e90e85cf`
* `ouroboros-network` (OPT): `a7fd4783a258ea10912ea2d6ea37a74103c0f5bf`

```
cabal run ouroboros-network:sim-tests -- -p '/progress.public root peers.progresses towards established peers/' +RTS -t -RTS --quickcheck-replay=824041
```

```
<<ghc: 71957914176 bytes, 13360 GCs, 13823445/141894912 avg/max bytes residency (35 samples), 367M in use, 0.001 INIT (0.000 elapsed), 12.301 MUT (12.268 elapsed), 1.873 GC (1.883 elapsed) :ghc>>
<<ghc: 71957923048 bytes, 13360 GCs, 14411011/140714216 avg/max bytes residency (35 samples), 371M in use, 0.001 INIT (0.000 elapsed), 12.067 MUT (12.038 elapsed), 1.864 GC (1.874 elapsed) :ghc>>
<<ghc: 71957932856 bytes, 13360 GCs, 12769673/120791008 avg/max bytes residency (33 samples), 282M in use, 0.001 INIT (0.001 elapsed), 12.241 MUT (12.222 elapsed), 1.711 GC (1.722 elapsed) :ghc>>
```

### Optimised IOSim, Unoptimised O-N

* `io-sim` (OPT): `3be6e9b8ffe8ba75ad88c7481cf846c9e90e85cf`
* `ouroboros-network` (UNOPT): `0880b1e061976729a277b73307927412ef8fac2c`

```
cabal run ouroboros-network:sim-tests -- -p '/progress.public root peers.progresses towards established peers/' +RTS -t -RTS --quickcheck-replay=824041
```

```
<<ghc: 71912404328 bytes, 13349 GCs, 10982478/91728608 avg/max bytes residency (27 samples), 232M in use, 0.000 INIT (0.000 elapsed), 11.449 MUT (11.426 elapsed), 1.432 GC (1.441 elapsed) :ghc>>
<<ghc: 71912384144 bytes, 13349 GCs, 10351128/94338568 avg/max bytes residency (27 samples), 226M in use, 0.001 INIT (0.000 elapsed), 11.939 MUT (11.919 elapsed), 1.449 GC (1.460 elapsed) :ghc>>
<<ghc: 71912403616 bytes, 13349 GCs, 10423612/89269200 avg/max bytes residency (27 samples), 231M in use, 0.000 INIT (0.000 elapsed), 11.846 MUT (11.821 elapsed), 1.453 GC (1.463 elapsed) :ghc>>
```

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs [hunk 1]
```diff
@@ -663,19 +664,21 @@ jobDemoteEstablishedPeer PeerSelectionActions{peerStateActions = PeerStateAction
...
+                                  = Set.size $ EstablishedPeers.toSet establishedPeers
+                                    `Set.intersection`
+                                    bigLedgerPeers
+        in
         Decision {
         decisionTrace = if peeraddr `Set.member` bigLedgerPeers
                         then [TraceDemoteWarmBigLedgerPeerFailed
                                targetNumberOfEstablishedBigLedgerPeers
-                               (Set.size $ EstablishedPeers.toSet establishedPeers
-                                           `Set.intersection`
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs:672]: Note that here is `intersection`

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs [hunk 2]
```diff
@@ -663,19 +664,21 @@ jobDemoteEstablishedPeer PeerSelectionActions{peerStateActions = PeerStateAction
...
-                                           bigLedgerPeers)
+                               numEstablishedBigLedgerPeers
                                peeraddr e]
                         else [TraceDemoteWarmFailed
                                targetNumberOfEstablishedPeers
-                               (Set.size $ EstablishedPeers.toSet establishedPeers
-                                    Set.\\ bigLedgerPeers)
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs:678]: And here is difference `\\`

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection.hs [hunk 3]
```diff
@@ -1726,7 +1726,7 @@ prop_governor_target_known_5_no_shrink_below env =
               --
               -- By subtracting a sum of `y` and `y'` we also do not account
               -- forgetting big ledger peers.
-              (\(x,y) (x',y') -> x Set.\\ x' Set.\\ (y <> y'))
+              (\(x,y) (x',y') -> x Set.\\ x' Set.\\ y Set.\\ y')
```
Comment (bolt12) [inline ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection.hs:1728]: Why this change? It seems the comment was aware of that sum

**Review by bolt12 — CHANGES_REQUESTED** [review]
LGTMm just a few remarks! Also, are the bangs everywhere really needed?

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection.hs (hunk 3 above)
Comment (coot) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection.hs:1728]: there's no difference, isn't it?

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs (hunk 1 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs:672]: Right, I'll remove this.

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection.hs (hunk 3 above)
Comment (bolt12) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/PeerSelection.hs:1728]: Right, I think that's right 😄

**Comment by coot** [comment]
> Also, are the bangs everywhere really needed?

I was thinking to remove the ones from `PeerSelectionCounters`, but they can save 10MB of heap in larger test cases.
